E106: Don't use relative paths for meta/main.yml

nre-ableton opened the pull request:
This change calculates the role's root directory by finding the
"tasks" directory first, rather than using the parent of the parent of
the current file. The problem with using the parent/parent scheme is
that roles which organize their tasks into subdirectories would cause
ansible-lint to erroneously raise E106 for an invalid role name.

---

Fixes https://github.com/ansible/ansible-lint/issues/1001

ssbarnea reviewed (changes requested):
AFAIK, roles can exist without tasks but they cannot exist without meta. Also, meta is always at predictable location, regardless how the tasks are organized.

If we are fixing an existing bug, we clearly need a test that prevents a future regression, a test that would reproduce the current bug.

It would even be better if you can add a test that reproduces the current bug and add it using @pytest.mark.xfail before making this fix.

I am saying this because I want to avoid future regressions, especially as this area was touched multiple times, and recently too.

nre-ableton commented:
Adding a task in a subdirectory of an existing role (such as `valid-due-to-meta`) reproduces the behavior. I tried to explain this in the commit message, but I think that your suggestion is much better. As such, I split the original commit to added the subdirectory and `@pytest.mark.xfail` in a separate commit with explanation there.

ssbarnea commented:
> Adding a task in a subdirectory of an existing role (such as `valid-due-to-meta`) will result in a failing test and reproduce the behavior. I tried to explain this in the commit message, but I think that your suggestion is much better. As such, I split the original commit to added the subdirectory and `@pytest.mark.xfail` in a separate commit with explanation there.

In that case maybe just add a comment on the empty task file explaining why the file is there, so we do not remove it by mistake?

nre-ableton commented:
> > Adding a task in a subdirectory of an existing role (such as `valid-due-to-meta`) will result in a failing test and reproduce the behavior. I tried to explain this in the commit message, but I think that your suggestion is much better. As such, I split the original commit to added the subdirectory and `@pytest.mark.xfail` in a separate commit with explanation there.
> 
> In that case maybe just add a comment on the empty task file explaining why the file is there, so we do not remove it by mistake?

Great suggestion, done! Also, I removed the `@pytest.mark.xfail` comment in the subsequent commit that fixes the problem, I forgot to do this before. :sweat_smile:

nre-ableton commented:
@ssbarnea Thanks for the quick review and merge! :clap:
